src/coreclr/jit/emit.cpp

```diff
@@ -5963,9 +5965,97 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
 
         emitCurIG = ig;
 
-        for (unsigned cnt = ig->igInsCnt; cnt; cnt--)
+        for (unsigned cnt = ig->igInsCnt; cnt > 0; cnt--)
         {
+#ifdef DEBUG
+            size_t     curInstrAddr = (size_t)cp;
+            instrDesc* curInstrDesc = id;
+#endif
+
             castto(id, BYTE*) += emitIssue1Instr(ig, id, &cp);
+
+#ifdef DEBUG
+            // Print the alignment boundary
+            if ((emitComp->opts.disAsm || emitComp->verbose) && (emitComp->opts.disAddr || emitComp->opts.disAlignment))
+            {
+                size_t      afterInstrAddr   = (size_t)cp;
+                instruction curIns           = curInstrDesc->idIns();
+                bool        isJccAffectedIns = false;
+
+#if defined(TARGET_XARCH)
+
+                // Determine if this instruction is part of a set that matches the Intel jcc erratum characteristic
+                // described here:
+                // https://www.intel.com/content/dam/support/us/en/documents/processors/mitigations-jump-conditional-code-erratum.pdf
+                // This is the case when a jump instruction crosses a 32-byte boundary, or ends on a 32-byte boundary.
+                // "Jump instruction" in this case includes conditional jump (jcc), macro-fused op-jcc (where 'op' is
+                // one of cmp, test, add, sub, and, inc, or dec), direct unconditional jump, indirect jump,
+                // direct/indirect call, and return.
+
+                size_t jccAlignBoundary     = 32;
```

Comment:
I have a couple questions about this comment.
1. We don't do anything about jcc erratum in adaptive loop alignment mode, currently. Are you suggesting we do?
2. I don't see how this code for `compJitAlignLoopForJcc` works. First, it's DEBUG only, so it appears to have been an experiment only, not in the shipping product. Also, how does incrementing `currentOffset` affect whether the jcc erratum condition is avoided?

Reply:
> We don't do anything about jcc erratum in adaptive loop alignment mode, currently. Are you suggesting we do?

In general, if we think that it degrades performance (as you observed), then yes, we should at least do it for instructions - Here is the graph from the findings made in https://github.com/dotnet/runtime/issues/35730:

Axis:
- X: Ratio of after/before. Stated another way, the ratio is (withmcu/withoutmcu). Ratios less than 1 mean the benchmark performed better with the JCC microcode update applied. Ratios greater than 1 mean the benchmark performed worse with the JCC microcode update applied.
- Y: Count of benchmarks in the bucket.

![80840714-ba0bda80-8bb2-11ea-8869-07ec55182661](https://user-images.githubusercontent.com/12488060/125116693-f170f380-e0a1-11eb-89d5-57d7611b63cd.png)

In general, I see more Microbenchmarks degraded (be it by small amount) than improved. 

My proposal would be to at least do JCC erratum for jumps that participate in the loop (like backedge).

> I don't see how this code for compJitAlignLoopForJcc works. First, it's DEBUG only, so it appears to have been an experiment only, not in the shipping product. Also, how does incrementing currentOffset affect whether the jcc erratum condition is avoided?

Yes, it was intentionally made for DEBUG only because it didn't fully solve the JCC erratum nor did it give big benefits. The way it works is - for non-adaptive alignment, if the loop already starts from an offset such that it will still fit in minimum no. of blocks required, then we skip aligning it. Before determining this, if `JitAlignLoopForJcc=1`, we just increase the offset from which the loop starts, so that the last backedge is pushed further down.  The assumption is, if there was JCC erratum, we would hope that our condition `if (currentOffset > extraBytesNotInLoop)` would do the right thing and not align the last backedge at the boundary. The more I think now - it won't work properly in most of the cases, and we need some more tracking to make sure that the backedges don't fall on the boundary. On the contrary, it could also happen that previously the backedge was not on boundary and after aligning the loop, it falls on the boundary, worsening the performance, and that too needs to be handle correctly.